compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/IrInlineCodegen.kt

```diff
@@ -194,6 +195,41 @@ class IrInlineCodegen(
             ::IrDefaultLambda
         )
     }
+
+    override fun inlineCall(
```

Comment:
Instead of hacking `IrInlineCodegen`, maybe create a separate implementation of `IrCallGenerator` that uses `ClassCodegen.generateMethodNode`?

Reply:
That's a really good idea, thanks. I've changed the code to use a different call generator.

One thing I'm a bit unclear on is whether we still need to replicate the inline cycle check or not. I'm thinking it shouldn't matter, since in the worst case we'll just put the throw instruction in a slightly different place. What do you think?

Reply:
Well, the only way we could get an error *here* is if an inline function with defaults calls itself, but direct recursion is already checked by the frontend way before that. You can see that because the error message is different:
```
x.kt:1:28: error: inline function 'public inline fun f(x: Int = ...): Unit defined in root package in file x.kt' cannot be recursive
inline fun f(x: Int = 1) { f() }
                           ^
```
vs
```
x.kt:1:28: error: the 'g' invocation is a part of inline cycle
inline fun f(x: Int = 1) { g() }
                           ^
x.kt:3:18: error: the 'f' invocation is a part of inline cycle
inline fun g() = f()
                 ^
```
With indirect recursion, the check would trigger when generating `f` or `g`, not `f$default`.

Reply:
On second though, if we start generating from `f`, then the cycle will indeed be detected at the call of `f` in `f$default`, and calling `generateMethodNode` may be fatal.

Reply:
Alright, it looks like this can't happen right now, but I've added the check anyway to be on the safe side. Thanks for the feedback @pyos!